src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -1800,20 +1812,32 @@ public ImmutableArray<DocumentId> GetDocumentIdsWithFilePath(string? filePath)
         private ImmutableDictionary<ProjectId, CompilationTracker> CreateCompilationTrackerMap(ProjectId projectId, ProjectDependencyGraph dependencyGraph)
         {
             var builder = ImmutableDictionary.CreateBuilder<ProjectId, CompilationTracker>();
-            var dependencies = dependencyGraph.GetProjectsThatTransitivelyDependOnThisProject(projectId);
+            IEnumerable<ProjectId>? dependencies = null;
 
-            foreach (var projectIdAndTracker in _projectIdToTrackerMap)
+            foreach (var (id, tracker) in _projectIdToTrackerMap)
             {
-                var id = projectIdAndTracker.Key;
-                var tracker = projectIdAndTracker.Value;
-
                 if (!tracker.HasCompilation)
                 {
                     continue;
                 }
 
-                var canReuse = id == projectId || !dependencies.Contains(id);
-                builder.Add(id, canReuse ? tracker : tracker.Fork(tracker.ProjectState));
+                builder.Add(id, CanReuse() ? tracker : tracker.Fork(tracker.ProjectState));
+
+                // Returns true if 'tracker' can be reused for project 'id'
+                bool CanReuse()
+                {
+                    if (id == projectId)
+                        return true;
+
+                    var forwardDependencies = dependencyGraph.TryGetProjectsThatThisProjectTransitivelyDependsOn(id);
+                    if (forwardDependencies is object && !forwardDependencies.Contains(projectId))
+                    {
+                        return true;
+                    }
+
+                    dependencies ??= dependencyGraph.GetProjectsThatTransitivelyDependOnThisProject(projectId);
+                    return !dependencies.Contains(projectId);
```

Comment:
would probably be good to document all these conditions.

Reply:
I added a comment explaining what the conditions check. I did not add a comment explaining why the condition is correct (such a comment was not available prior to my change so I don't have the information).